### PR TITLE
Calculation of NLP properties with batches

### PR DESCRIPTION
**Pull request by ItayGabbay**
Before/after time comparison for 3000 samples:

| Batch size | Time(s) |
|--------|--------|
| 1 (no batches) | 124 |
| 16 | **12** |
| 32 | **5** | 

In addition, I added handling of exceptions in case the batch_size is too large - we're gradually reducing the batch size until there is sufficient memory.

### CODE REVIEW

File: deepchecks/nlp/utils/text_properties.py [hunk 1]
```diff
@@ -870,44 +900,47 @@ def calculate_builtin_properties(
...
-        unit='Text Sample'
-    )
-    for text in progress_bar:
-        progress_bar.set_postfix(
-            {'Sample': truncate_string(text, max_length=20) if text else 'EMPTY STRING'},
-            refresh=False
-        )
-        if pd.isna(text):
+    with tqdm(total=int(len(raw_text)//batch_size)) as pbar:
+        for i in range(0, len(raw_text), batch_size):
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:906]: you can do tqdm(range(0, len(raw_text), batch_size)) than you dont need to manually update pbar

File: deepchecks/nlp/utils/text_properties.py [hunk 2]
```diff
@@ -870,44 +900,47 @@ def calculate_builtin_properties(
...
+            batch = raw_text[i:i + batch_size]
+            batch_properties = defaultdict(list)
+            pbar.update(1)
+
+            # filtering out empty sequences
+            filtered_sequences = [seq for seq in batch if pd.isna(seq) is False]
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:912]: what do you do with them? return none in the relevant places?

**Review by yromanyshyn — APPROVED** [review]

File: deepchecks/nlp/utils/text_properties.py (hunk 2 above)
Comment (ItayGabbay) [reply deepchecks/nlp/utils/text_properties.py:912]: np.nan. as it was before

File: deepchecks/nlp/utils/text_properties.py [hunk 3]
```diff
@@ -374,28 +380,53 @@ def _predict(text: str, classifier, kind: str) -> float:
...
+                elif kind == 'fluency':
+                    results.append(v['score'] if v['label'] == 'LABEL_1' else 1 - v['score'])
+                elif kind == 'formality':
+                    results.append(v['score'] if v['label'] == 'formal' else 1 - v['score'])
+                else:
+                    raise ValueError('Unsupported value for "kind" parameter')
+
+            return results  # Return the results if prediction is successful
+
+        except Exception:  # pylint: disable=broad-except
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:425]: seems that for a batch of size 32 that has one bad example you will always come the at the end to running everything sample by sample but after you run many unnecessary calculations

File: deepchecks/nlp/utils/text_properties.py [hunk 4]
```diff
@@ -870,40 +917,41 @@ def calculate_builtin_properties(
...
-            for prop in text_properties:
-                calculated_properties[prop['name']].append(np.nan)
-            continue
-        sample_language = run_available_kwargs(language, text=text, **kwargs)
+    for i in tqdm(range(0, len(raw_text), batch_size)):
+        batch = raw_text[i:i + batch_size]
+        batch_properties = defaultdict(list)
+
+        # filtering out empty sequences
+        filtered_sequences = [seq for seq in batch if pd.isna(seq) is False]
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:925]: (if i understood currently) You are filtering out stuff but dont return none in the relevant places. if i send you 100 samples and you give me back a dataframe with 93 values, how can i work with this?

**Review by Nadav-Barak — CHANGES_REQUESTED** [review]
This is a big functionality, we should have several tests for it

File: deepchecks/nlp/utils/text_properties.py [hunk 5]
```diff
@@ -870,40 +917,41 @@ def calculate_builtin_properties(
...
+                    if prop['name'] in BATCH_PROPERTIES:
+                        value = run_available_kwargs(func=prop['method'], text_batch=filtered_sequences, **kwargs)
+                    else:
+                        value = _batch_wrapper(text_batch=filtered_sequences, func=prop['method'], **kwargs)
+                    batch_properties[prop['name']].extend(value)
                 except ImportError as e:
                     warnings.warn(warning_message.format(prop['name'], str(e)))
-                    calculated_properties[prop['name']].append(np.nan)
+                    batch_properties[prop['name']].extend([np.nan] * len(batch))
                     import_warnings.add(prop['name'])
```
Comment (nirhutnik) [inline deepchecks/nlp/utils/text_properties.py:950]: Not sure I followed here. I don't see how you keep the order here. what if I have 32 samples, and sample number 7 is not in English? Where do you make sure that number 7 gets the "none" value? I only see that you filter by language but don't fill the others with None, and make sure to keep the order.

Comment (ItayGabbay) [reply deepchecks/nlp/utils/text_properties.py:961]: Fixed.

File: deepchecks/nlp/utils/text_properties.py (hunk 4 above)
Comment (ItayGabbay) [reply deepchecks/nlp/utils/text_properties.py]: Fixed.

File: deepchecks/nlp/utils/text_properties.py (hunk 3 above)
Comment (ItayGabbay) [reply deepchecks/nlp/utils/text_properties.py:426]: Added a retry mechanism to avoid that

File: deepchecks/nlp/utils/text_properties.py [hunk 6]
```diff
@@ -374,28 +381,54 @@ def _predict(text: str, classifier, kind: str) -> float:
...
+                if not v:
+                    results.append(np.nan)
+                elif kind == 'toxicity':
+                    results.append(v['score'])
+                elif kind == 'fluency':
+                    results.append(v['score'] if v['label'] == 'LABEL_1' else 1 - v['score'])
+                elif kind == 'formality':
+                    results.append(v['score'] if v['label'] == 'formal' else 1 - v['score'])
+                else:
+                    raise ValueError('Unsupported value for "kind" parameter')
```
Comment (Nadav-Barak) [inline deepchecks/nlp/utils/text_properties.py:422]: Duplicate code can copy to function

File: tests/nlp/utils/test_properties.py [hunk 7]
```diff
@@ -182,6 +182,25 @@ def test_calculate_average_sentence_length_property(tweet_emotion_train_test_tex
...
+
+    # Act
+    for batch in batch_sizes:
+        result = calculate_builtin_properties(test_text, include_properties=['Average Words Per Sentence'],
+                                              batch_size=batch)[0]
+        result_none_text = calculate_builtin_properties([None], include_properties=['Average Words Per Sentence'],
+                                                        batch_size=batch)[0]
+
+        # Assert
+        assert_that(result['Average Words Per Sentence'][0: 10], equal_to([5.667, 7.0, 11.0, 12.0, 8.0, 19.0, 3.0, 9.0,
```
Comment (Nadav-Barak) [inline tests/nlp/utils/test_properties.py:199]: add assertion also on property that is english only to test if works well also there

**Review by Nadav-Barak — APPROVED** [review]

File: deepchecks/nlp/utils/text_properties.py [hunk 8]
```diff
@@ -374,28 +381,54 @@ def _predict(text: str, classifier, kind: str) -> float:
...
+                            raise ValueError('Unsupported value for "kind" parameter')
+                    except Exception:  # pylint: disable=broad-except
+                        results.append(np.nan)
+                return results  # Return the results if prediction is successful
+
+            v_list = classifier(text_list_to_predict, batch_size=reduced_batch_size)
+            results = []
+
+            for v in v_list:
+                if not v:
```
Comment (noamzbr) [inline deepchecks/nlp/utils/text_properties.py:413]: seems like there's easily avoidable code duplication. Maybe take this whole switch clause to an external function?

File: deepchecks/nlp/utils/text_properties.py [hunk 9]
```diff
@@ -614,6 +650,19 @@ def average_syllable_length(text: str, cmudict_dict: dict = None) -> float:
...
 
 
+def _batch_wrapper(text_batch: Sequence[str], func: Callable, **kwargs) -> List[Any]:
+    """Wrap the non-batched properties execution with batches API."""
+    results = []
+    language_property_result = []
+    if 'language_property_result' in kwargs:
+        language_property_result = kwargs.pop('language_property_result')
+    for i, text in enumerate(text_batch):
+        kwargs['language_property_result'] = language_property_result[i] if i < len(language_property_result) else None
```
Comment (noamzbr) [inline deepchecks/nlp/utils/text_properties.py:660]: didn't understand this line

File: deepchecks/nlp/utils/text_properties.py [hunk 10]
```diff
@@ -870,40 +925,54 @@ def calculate_builtin_properties(
...
+                        and ignore_non_english_samples_for_english_properties is True:
+                    filtered_sequences = [e for i, e in enumerate(filtered_sequences) if i not in non_english_indices]
+                kwargs['batch_size'] = batch_size
                 try:
-                    value = run_available_kwargs(prop['method'], text=text, **kwargs)
-                    calculated_properties[prop['name']].append(value)
+                    if prop['name'] in BATCH_PROPERTIES:
+                        value = run_available_kwargs(func=prop['method'], text_batch=filtered_sequences, **kwargs)
+                    else:
+                        value = _batch_wrapper(text_batch=filtered_sequences, func=prop['method'], **kwargs)
```
Comment (noamzbr) [inline deepchecks/nlp/utils/text_properties.py:958]: I kinda hate this, can we perhaps add _batch_wrapper inside the original property dict? It's ugly duplication but I don't like that we have these gazillion lists of property types...

Comment (ItayGabbay) [reply deepchecks/nlp/utils/text_properties.py:958]: I think it makes sense to have batched & non-batched properties

**Pull request merged**